### PR TITLE
All DateFormatters should be able to print Long.MAX_VALUE

### LINKED ISSUE #65639: Failed to index event containing name root
**Elasticsearch version** (`bin/elasticsearch --version`):  7.9.2

**Plugins installed**: []

**JVM version** (`java -version`): 15+36-1562

**OS version** (`uname -a` if on a Unix-like system): 5.6.13

**Description of the problem including expected versus actual behavior**:

Sending audit logs to Logstash 7.10 over syslog tcp which is indexing events to elasticsearch. When any audit event contains the [user][name] value of "root" I receive the error below. I'm receiving events that do not contain the [user][name] of root. This is an index created with the ECS-1.4 template

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem,
including (e.g.) index creation, mappings, settings, query etc.  The easier
you make for us to reproduce it, the more likely that somebody will take the
time to look at it.

 1. Import ECS-1.4 template from https://github.com/elastic/ecs
 2. Index a event with [user][name] : root
 3.

**Provide logs (if relevant)**:

```[2020-11-30T17:50:24,828][WARN ][logstash.outputs.elasticsearch][vmware][xxxxx] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"ecs-1.4", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x502be27b>], :response=>{"index"=>{"_index"=>"ecs-1.4-000006", "_type"=>"_doc", "_id"=>"xxxxxxxxx", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"failed to parse field [user.name] of type [keyword] in document with id 'xxxxxxxxx'. Preview of field's value: '{name=root}'", "caused_by"=>{"type"=>"illegal_state_exception", "reason"=>"Can't get text on a START_OBJECT at 1:44"}}}}}```

### LINKED ISSUE #65760: java.time.DateTimeException when sorting by nullable date and requested size equals total hits
**Elasticsearch version**: 7.6 — 7.10

**JVM version**:
```
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-b04)
OpenJDK 64-Bit Server VM (build 25.212-b04, mixed mode)
```
**OS version**: Linux

**Description of the problem including expected versus actual behavior**:

I have an index with a nullable date field. When I query items sorted by the field and specify size which equals total hits count, I get error:
```json
{
    "type": "response_handler_failure_transport_exception",
    "reason": "java.time.DateTimeException: Field Year cannot be printed as the value 292278994 exceeds the maximum print width of 4"
}
```

**Steps to reproduce**:
1. Create index:
```json
PUT /bug-reproduce

{
  "mappings": {
        "properties": {
          "date": {
            "type": "date",
            "format": "basic_date_time_no_millis"
          }
        }
    },
  "settings": {
    "index": {
      "number_of_shards": "1",
      "number_of_replicas": "0"
     }
  }
}
```
2. Add 2 documents:
```
PUT /bug-reproduce/_doc/1

{
    "date": "20201202T190000Z"
}

PUT /bug-reproduce/_doc/2

{
    "date": null
}
```
3. Try searching:
```json
POST /bug-reproduce/_search

{
  "size": 2,
  "sort": [
    {
      "date": {}
    }
  ]
}
```
4. Get error:
```json
{
  "error":{
    "root_cause":[
      {
        "type":"response_handler_failure_transport_exception",
        "reason":"java.time.DateTimeException: Field Year cannot be printed as the value 292278994 exceeds the maximum print width of 4"
      }
    ],
    "type":"search_phase_execution_exception",
    "reason":"all shards failed",
    "phase":"query",
    "grouped":true,
    "failed_shards":[
      {
        "shard":0,
        "index":"bug-reproduce",
        "node":"nyQL21NNT2KkEcJCEahP1Q",
        "reason":{
          "type":"response_handler_failure_transport_exception",
          "reason":"java.time.DateTimeException: Field Year cannot be printed as the value 292278994 exceeds the maximum print width of 4",
          "caused_by":{
            "type":"date_time_exception",
            "reason":"Field Year cannot be printed as the value 292278994 exceeds the maximum print width of 4"
          }
        }
      }
    ]
  },
  "status":500
}
```
If I specify `size: 1` or `size: 3` there is no error

### PR DESCRIPTION
**Pull request by cbuescher**
All date formats should allow all possible long date values to be formatted.
Currently some of the printers used by some formats are too restrictive here
(e.g. for 'basic_date_time_no_millis'), only allowing up to four digits for the
year. This PR changes this for the formatters that are otherwise failing the
added tests that tries to format Long.MAX_VALUE and Long.MIN_VALUE respectively,
as these are values we for example implicitely use when sorting missing fields.

Closes #65760

### CODE REVIEW

**Comment by cbuescher** [comment]
@pgomulka since we already talked about this, would you mind taking a first look? I'm not too sure which side effects these changes might have, so please suggest an alternative approach if there is a better one.

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 1]
```diff
@@ -466,7 +466,7 @@
 
     private static final DateTimeFormatter STRICT_BASIC_WEEK_DATE_PRINTER = new DateTimeFormatterBuilder()
         .parseStrict()
-        .appendValue(IsoFields.WEEK_BASED_YEAR, 4)
+        .appendValue(IsoFields.WEEK_BASED_YEAR, 4, 10, SignStyle.NEVER)
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatters.java:469]: why SignStyle.NEVER?
I guess we could try with NORMAL (this is default)

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 2]
```diff
@@ -345,7 +345,7 @@
     );
 
     private static final DateTimeFormatter BASIC_YEAR_MONTH_DAY_FORMATTER = new DateTimeFormatterBuilder()
-        .appendValue(ChronoField.YEAR, 4, 4, SignStyle.NORMAL)
+        .appendValue(ChronoField.YEAR, 4, 10, SignStyle.NORMAL)
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatters.java:348]: I think we should have a separate BASIC_YEAR_MONTH_DAY_PRINTER
FORMATTER is a bit unfortunate name to be fair, as it is used for parsing..

**Review by pgomulka — COMMENTED** [review]
I left two comments, and have changes to fix them

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java (hunk 1 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java:469]: Sure, I think I took this from elsewhere but NORMAL sounds right to me.

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java (hunk 2 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java:348]: 👍

**Comment by cbuescher** [comment]
@pgomulka thanks for the review, I pushed some changes.

**Comment by cbuescher** [comment]
@elasticmachine update branch

**Review by pgomulka — APPROVED** [review]
LGTM

**Pull request merged**